crates/store/src/state/state_db.rs

```diff
@@ -1,60 +1,257 @@
+        use JournalEntry::*;
+
+        match self {
+            UpdateRaw { key, prev_value } => {
+                state_db.dirty_state.insert(key, prev_value);
+            }
+            SetAccountCount { prev_count } => {
+                state_db.dirty_account_count = Some(prev_count);
+            }
+            InsertScript {
+                script_hash,
+                prev_exist,
+            } => {
+                if !prev_exist {
+                    state_db.dirty_scripts.remove(&script_hash);
+                }
+            }
+            InsertData {
+                data_hash,
+                prev_exist,
+            } => {
+                if !prev_exist {
+                    state_db.dirty_data.remove(&data_hash);
+                }
+            }
+            AppendLog { index } => {
+                assert_eq!(state_db.dirty_logs.len(), index + 1);
+                state_db.dirty_logs.pop_back();
+            }
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct Journal {
+    entries: im::Vector<JournalEntry>,
+}
+
+impl Journal {
+    pub fn len(&self) -> usize {
+        self.entries.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.entries.is_empty()
+    }
+
+    /// Revert journal to len
+    /// Return reverted entries sorted in journal emit order
+    pub fn revert_entries(&mut self, len: usize) -> im::Vector<JournalEntry> {
+        self.entries.slice(len..)
+    }
+
+    pub fn push(&mut self, entry: JournalEntry) {
+        self.entries.push_back(entry);
+    }
+
+    pub fn clear(&mut self) {
+        self.entries.clear();
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Revision {
+    id: usize,
+    journal_len: usize,
+}
+
+#[derive(Debug, Default)]
+pub struct StateTracker {
+    touched_keys: Mutex<HashSet<H256>>,
+    write_data: Mutex<HashMap<H256, Bytes>>,
+    read_data: Mutex<HashMap<H256, Bytes>>,
+}
+
+impl StateTracker {
+    /// Return touched keys
+    pub fn touched_keys(&self) -> &Mutex<HashSet<H256>> {
+        &self.touched_keys
+    }
+
+    pub fn write_data(&self) -> &Mutex<HashMap<H256, Bytes>> {
+        &self.write_data
+    }
+
+    pub fn read_data(&self) -> &Mutex<HashMap<H256, Bytes>> {
+        &self.read_data
+    }
+
+    /// Record a key in the tracker
+    pub fn touch_key(&self, key: &H256) {
+        self.touched_keys.lock().unwrap().insert(*key);
+    }
 }
 
-pub struct StateTree<'a> {
-    tree: SMT<SMTStore<'a, StoreTransaction>>,
-    account_count: u32,
-    context: StateContext,
+pub struct StateDB<S> {
+    /// inner state
+    state: S,
+    /// journals
+    journal: Journal,
+    /// next_revision_id
+    next_revision_id: usize,
+    /// revisions
+    revisions: im::Vector<Revision>,
+    /// dirty state in memory
+    dirty_state: im::HashMap<H256, H256>,
+    /// dirty account count
+    dirty_account_count: Option<u32>,
+    /// dirty scripts
+    dirty_scripts: im::HashMap<H256, packed::Script>,
```

Comment:
I'm wondering whether we should use regular `HashMap`/`Vec` instead of immutable ones if we are using journals. Do we really need fast cloning of a dirty `StateDB`?